Zipkin UI submodule

eirslett opened the pull request:
I tried moving the UI resources into their own submodule, separate from zipkin-web.
@abesto do you have time to look at the gradle stuff? I'm a complete Gradle n00b, so I just copy-pasted stuff around ;-)
If you run gradlew zipkin-ui:jar, it runs webpack/karma and builds a jar that only consists of assets.
I changed the layout a bit, so css/js isn't served from a /dist folder, it's served directly from root.

What we need to do now, is somehow import these resources into zipkin-web and make the Scala project serve them. I think we need a minor configuration tweak in the Scala code, so it serves assets from / instead of /dist.

Then, we need to publish zipkin-ui to a repository, so that zipkin-java can use it too.
I'm not sure .jar is the best container format, if possible it would be nice to package as .zip instead.

eirslett committed:
move zipkin ui into its own submodule

eirslett committed:
Tweak zipkin ui settings for gradle etc.

This is an attempt to set up a node/npm-like
directory structure that is understood by Gradle.
It builds an artifact that consists of only static
resources, and is intended to be imported into
the other zipkin projects as a dependency.

abesto commented:
Looks reasonable at a glance, I'll do a proper review tonight. Thanks!

codefromthecrypt commented:
great. I like this mission

abesto commented on zipkin-ui/build.gradle:
```diff
@@ -0,0 +1,45 @@
+build.dependsOn(webpack)
+build.dependsOn(karma)
+processResources.dependsOn(webpack)
+processResources.dependsOn(karma)
```
Here's how I'd set up the task dependencies. `check` is the Gradle standard "run all the tests and whatevers" task; `build` depends on it. I'm assuming the `karma` tests need the `webpack` build to have completed.

```
npmInstall ---> webpack ---> karma ---> check
                               |
                               + -----> processResources
```

This includes (either explicitly or transitively) all the dependencies you've specified, except for one: `karma -> processResources`. I'm guessing there's no actual dependency there, can you verify that? If there is, we can just move `processResources` to depend on `karma` instead of `webpack` (which depends on `webpack`, so we still get it transitively)

abesto commented:
Other than the note on the dependency chain, this looks pretty good to me. I'm taking a quick look to see if there's an easy way to make web load assets from ui. Will have more time tomorrow, and then there's a long weekend coming, you know what that means :)

On the package format, I don't know what would be the relevant best practice in this case, or even which stack to draw it from. Maybe a war file, or maybe just a zip to keep it simple, but then a jar is just a glorified zip so... Yeah.

abesto commented:
Result of the short experiment: just adding a `compile project(':zipkin-ui')` dependency to `zipkin-web` gives `zipkin-web` access to the resources of `zipkin-ui`, but there's much confusion around paths, especially since in the current setup `index.html` is in the same directory in the jar as all the static assets. My current best idea is to organize the jar in a way that there's clear separation between html files (especially `index.html` which needs special treatment in the handlers), and static assets. For example something like

```
zipkin-ui.jar
  |
  +-- html
  |    |
  |    +-- *.html
  +-- assets
       |
       +-- *.css
       +-- *.js
```

Another option is to put just `index.html` in a special place, and let the static handler serve all other HTML files from the same source as all other static assets, adding `html => text/html` to the type map.

Another option is redesigning the way the handlers are structured, such that there's no need for special-casing `index.html`.

Am I making sense? It's getting late, and this is the first time I'm messing around with Finagle.

PS. Don't forget to remove the node, webpack, and karma stuff from `zipkin-web/build.gradle`

eirslett commented:
I think it's easier if they're in the same directory (the html/css/js files), but nothing's set in stone, we can configure that in webpack very easily. They're just in the same folder for simplicity - if simplicity isn't achieved, then that is pointless.
I think we should try to mimic apache behavior in zipkin-web, that it serves every file in the folder normally, and on / it serves index.html. (So maybe something like an "alias" for the route handler?)

BTW, there are no other html files, index.html will be the only html file that is ever served. It's not even a bit HTML file, just enough boilerplate to be able to import the css/javascript code and bootstrap the application. Even if you go to a page like /dependencies, you will still get index.html from the server, but the javascript UI will look at `window.location` and figure out what to render based on that.

abesto committed:
zipkin-web serves static assets from zipkin-ui

abesto committed:
[gradle] zipkin-ui: reorganize task dependencies

abesto committed:
zipkin-web: Serve index.html by default

This serves index.html for /dependency/ for example, which is the
intended behavior.

abesto commented:
Implemented the "redesign handler structure" option, with extra trickery to always serve `index.html` when in doubt.

abesto committed:
[gradle] zipkin-ui: remove another unneeded dependsOn

build depends on processResources, so it's enough if
processResources depends on webpack. No need for build to depend
on webpack directly as well.

codefromthecrypt commented:
lemme know if you two want third eyes on this, otherwise, merge when you think it is ready!

eirslett commented:
LGTM!
Let's just distribute it as a .jar for now, we could always change it later.

eirslett commented:
Hmm, i get a weird error:

```
Unknown error


com.twitter.zipkin.web.Handlers$$anon$1.com$twitter$zipkin$web$Handlers$$anon$$getStream(Handlers.scala:100)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1$$anonfun$apply$6$$anonfun$apply$8.apply(Handlers.scala:112)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1$$anonfun$apply$6$$anonfun$apply$8.apply(Handlers.scala:111)
scala.Option.flatMap(Option.scala:171)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1$$anonfun$apply$6.apply(Handlers.scala:111)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1$$anonfun$apply$6.apply(Handlers.scala:111)
scala.Option.orElse(Option.scala:289)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1.apply(Handlers.scala:110)
com.twitter.zipkin.web.Handlers$$anon$1$$anonfun$getRenderer$1.apply(Handlers.scala:109)
scala.Option.orElse(Option.scala:289)
com.twitter.zipkin.web.Handlers$$anon$1.getRenderer(Handlers.scala:108)
com.twitter.zipkin.web.Handlers$$anon$1.apply(Handlers.scala:126)
com.twitter.zipkin.web.Handlers$$anon$1.apply(Handlers.scala:95)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:13)
com.twitter.zipkin.web.Handlers$$anonfun$1.apply(Handlers.scala:67)
com.twitter.zipkin.web.Handlers$$anonfun$1.apply(Handlers.scala:66)
com.twitter.finagle.Filter$$anon$6.apply(Filter.scala:206)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:13)
com.twitter.zipkin.web.Handlers$$anonfun$4.apply(Handlers.scala:76)
com.twitter.zipkin.web.Handlers$$anonfun$4.apply(Handlers.scala:75)
com.twitter.finagle.Filter$$anon$6.apply(Filter.scala:206)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:13)
com.twitter.zipkin.web.Handlers$$anonfun$collectStats$1$$anonfun$apply$3.apply(Handlers.scala:60)
com.twitter.zipkin.web.Handlers$$anonfun$collectStats$1$$anonfun$apply$3.apply(Handlers.scala:60)
com.twitter.finagle.stats.Stat$.timeFuture(Stat.scala:39)
com.twitter.finagle.stats.Stat$.timeFuture(Stat.scala:48)
com.twitter.zipkin.web.Handlers$$anonfun$collectStats$1.apply(Handlers.scala:60)
com.twitter.zipkin.web.Handlers$$anonfun$collectStats$1.apply(Handlers.scala:59)
com.twitter.finagle.Filter$$anon$6.apply(Filter.scala:206)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.http.HttpMuxer.apply(HttpMuxer.scala:62)
com.twitter.finagle.http.HttpMuxer.apply(HttpMuxer.scala:25)
com.twitter.finagle.ServiceProxy.apply(Service.scala:120)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.tracing.ServerDestTracingProxy$$anon$1.apply(DestinationTracing.scala:32)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.service.DeadlineFilter.apply(DeadlineFilter.scala:168)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.filter.DtabStatsFilter.apply(DtabStatsFilter.scala:37)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.service.StatsFilter.apply(StatsFilter.scala:146)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.filter.ExceptionSourceFilter.apply(ExceptionSourceFilter.scala:33)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.filter.MkJvmFilter$$anon$1.apply(JvmFilter.scala:29)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.filter.ServerStatsFilter.apply(ServerStatsFilter.scala:50)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.tracing.AnnotatingTracingFilter.apply(TraceInitializerFilter.scala:159)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:13)
com.twitter.zipkin.web.FilteredHttpEntrypointTraceInitializer$$anonfun$com$twitter$zipkin$web$FilteredHttpEntrypointTraceInitializer$$newRootSpan$1.apply(FilteredHttpEntrypointTraceInitializer.scala:39)
com.twitter.zipkin.web.FilteredHttpEntrypointTraceInitializer$$anonfun$com$twitter$zipkin$web$FilteredHttpEntrypointTraceInitializer$$newRootSpan$1.apply(FilteredHttpEntrypointTraceInitializer.scala:32)
com.twitter.util.Local.let(Local.scala:141)
com.twitter.finagle.context.Context$class.let(Context.scala:176)
com.twitter.finagle.context.MarshalledContext.let(Context.scala:232)
com.twitter.finagle.tracing.Trace$$anonfun$letTracerAndId$1.apply(Trace.scala:216)
com.twitter.util.Local.let(Local.scala:141)
com.twitter.finagle.context.Context$class.let(Context.scala:176)
com.twitter.finagle.context.LocalContext.let(LocalContext.scala:8)
com.twitter.finagle.tracing.Trace$.letTracerAndId(Trace.scala:215)
com.twitter.zipkin.web.FilteredHttpEntrypointTraceInitializer$.com$twitter$zipkin$web$FilteredHttpEntrypointTraceInitializer$$newRootSpan(FilteredHttpEntrypointTraceInitializer.scala:32)
com.twitter.zipkin.web.FilteredHttpEntrypointTraceInitializer$$anonfun$1.apply(FilteredHttpEntrypointTraceInitializer.scala:22)
com.twitter.zipkin.web.FilteredHttpEntrypointTraceInitializer$$anonfun$1.apply(FilteredHttpEntrypointTraceInitializer.scala:20)
com.twitter.finagle.Filter$$anon$6.apply(Filter.scala:206)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.filter.MonitorFilter$$anonfun$apply$1.apply(MonitorFilter.scala:35)
com.twitter.finagle.filter.MonitorFilter$$anonfun$apply$1.apply(MonitorFilter.scala:35)
com.twitter.util.Future$$anonfun$monitored$1.apply$mcV$sp(Future.scala:121)
com.twitter.util.Monitor$$anonfun$apply$1.apply$mcV$sp(Monitor.scala:38)
com.twitter.util.Monitor$$anonfun$apply$1.apply(Monitor.scala:38)
com.twitter.util.Monitor$$anonfun$apply$1.apply(Monitor.scala:38)
com.twitter.util.Monitor$$anonfun$using$1.apply(Monitor.scala:110)
com.twitter.util.Monitor$.restoring(Monitor.scala:117)
com.twitter.util.Monitor$.using(Monitor.scala:108)
com.twitter.util.Monitor$class.apply(Monitor.scala:37)
com.twitter.util.Future$Monitored.apply(Future.scala:128)
com.twitter.util.Future$.monitored(Future.scala:120)
com.twitter.finagle.filter.MonitorFilter.apply(MonitorFilter.scala:34)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.http.filter.ServerContextFilter$$anonfun$apply$1.apply(ContextFilter.scala:17)
com.twitter.finagle.http.filter.ServerContextFilter$$anonfun$apply$1.apply(ContextFilter.scala:17)
com.twitter.finagle.http.codec.HttpContext$.read(HttpContext.scala:44)
com.twitter.finagle.http.filter.ServerContextFilter.apply(ContextFilter.scala:17)
com.twitter.finagle.http.filter.ServerContextFilter.apply(ContextFilter.scala:13)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.Service$$anon$2.apply(Service.scala:16)
com.twitter.finagle.http.filter.DtabFilter.apply(DtabFilter.scala:22)
com.twitter.finagle.http.filter.DtabFilter.apply(DtabFilter.scala:11)
com.twitter.finagle.Filter$$anon$1.apply(Filter.scala:70)
com.twitter.finagle.http.codec.HttpServerDispatcher.dispatch(HttpServerDispatcher.scala:73)
com.twitter.finagle.dispatch.GenSerialServerDispatcher$$anonfun$com$twitter$finagle$dispatch$GenSerialServerDispatcher$$loop$1.apply(ServerDispatcher.scala:51)
com.twitter.finagle.dispatch.GenSerialServerDispatcher$$anonfun$com$twitter$finagle$dispatch$GenSerialServerDispatcher$$loop$1.apply(ServerDispatcher.scala:45)
com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:921)
com.twitter.util.Future$$anonfun$flatMap$1.apply(Future.scala:920)
com.twitter.util.Promise$Transformer.liftedTree1$1(Promise.scala:112)
com.twitter.util.Promise$Transformer.k(Promise.scala:112)
com.twitter.util.Promise$Transformer.apply(Promise.scala:122)
com.twitter.util.Promise$Transformer.apply(Promise.scala:103)
com.twitter.util.Promise$$anon$1.run(Promise.scala:381)
com.twitter.concurrent.LocalScheduler$Activation.run(Scheduler.scala:178)
com.twitter.concurrent.LocalScheduler$Activation.submit(Scheduler.scala:136)
com.twitter.concurrent.LocalScheduler.submit(Scheduler.scala:207)
com.twitter.concurrent.Scheduler$.submit(Scheduler.scala:92)
com.twitter.util.Promise.runq(Promise.scala:350)
com.twitter.util.Promise.updateIfEmpty(Promise.scala:721)
com.twitter.util.Promise.update(Promise.scala:694)
com.twitter.util.Promise.setValue(Promise.scala:670)
com.twitter.concurrent.AsyncQueue.offer(AsyncQueue.scala:111)
com.twitter.finagle.netty3.transport.ChannelTransport.handleUpstream(ChannelTransport.scala:55)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
com.twitter.finagle.netty3.channel.ChannelRequestStatsHandler.messageReceived(ChannelRequestStatsHandler.scala:32)
org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:145)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.messageReceived(SimpleChannelUpstreamHandler.java:124)
com.twitter.finagle.http.codec.RespondToExpectContinue.messageReceived(RespondToExpectContinue.scala:30)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.messageReceived(SimpleChannelUpstreamHandler.java:124)
com.twitter.finagle.http.codec.PayloadSizeHandler.messageReceived(PayloadSizeHandler.scala:23)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.handler.codec.http.HttpContentEncoder.messageReceived(HttpContentEncoder.java:82)
org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459)
org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
org.jboss.netty.handler.codec.http.HttpServerCodec.handleUpstream(HttpServerCodec.java:56)
com.twitter.finagle.http.SafeHttpServerCodec.handleUpstream(Codec.scala:36)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142)
com.twitter.finagle.netty3.channel.ChannelStatsHandler.messageReceived(ChannelStatsHandler.scala:68)
org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
java.lang.Thread.run(Unknown Source)
```

This is on the / page, when running `gradlew zipkin-web:shadowJar` and then `java -jar` on that artifact...

abesto commented:
I consistently fail to check how things work with shadowJar (and am
consistently saddened that they don't work the same as when run from
Gradle). Will take a look tomorrow.

On Thu, Mar 10, 2016 at 4:06 PM Adrian Cole notifications@github.com
wrote:

> great. I like this mission
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1024#issuecomment-194890978.
> 
> ## 
> 
> Zoltán Nagy
> https://abesto.net

codefromthecrypt commented:
@eirslett @abesto weird error fixed https://github.com/openzipkin/zipkin/commit/284cbbf7989fc4e96695966a6fb6b449968026b2

abesto commented:
Awesome, thank you!

On Thu, Mar 10, 2016 at 4:06 PM Adrian Cole notifications@github.com
wrote:

> great. I like this mission
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1024#issuecomment-194890978.
> 
> ## 
> 
> Zoltán Nagy
> https://abesto.net
